### PR TITLE
HDDS-1553. Add metrics in rack aware container placement policy.

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-1553.

Initial patch

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 59 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 68 | Maven dependency ordering for branch |
| +1 | mvninstall | 719 | trunk passed |
| +1 | compile | 443 | trunk passed |
| +1 | checkstyle | 77 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 861 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 173 | trunk passed |
| 0 | spotbugs | 435 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 638 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 36 | Maven dependency ordering for patch |
| +1 | mvninstall | 567 | the patch passed |
| +1 | compile | 398 | the patch passed |
| +1 | javac | 398 | the patch passed |
| -0 | checkstyle | 40 | hadoop-hdds: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 763 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 189 | the patch passed |
| +1 | findbugs | 776 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 343 | hadoop-hdds in the patch failed. |
| -1 | unit | 1345 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 7761 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.container.placement.algorithms.TestSCMContainerPlacementRackAware |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.ozone.client.rpc.TestContainerStateMachine |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1361 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b28d7db4c212 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 55cc115 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/testReport/ |
| Max. process+thread count | 1537 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 75 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 71 | Maven dependency ordering for branch |
| +1 | mvninstall | 643 | trunk passed |
| +1 | compile | 362 | trunk passed |
| +1 | checkstyle | 71 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 933 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 170 | trunk passed |
| 0 | spotbugs | 421 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 628 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for patch |
| +1 | mvninstall | 549 | the patch passed |
| +1 | compile | 378 | the patch passed |
| +1 | javac | 378 | the patch passed |
| +1 | checkstyle | 77 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 757 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 173 | the patch passed |
| +1 | findbugs | 656 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 356 | hadoop-hdds in the patch passed. |
| -1 | unit | 2379 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 8523 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.TestOzoneConfigurationFields |
|   | hadoop.ozone.client.rpc.TestOzoneClientRetriesOnException |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1361 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 1fe3ca2fca57 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 371c9eb |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/2/testReport/ |
| Max. process+thread count | 5092 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 64 | Maven dependency ordering for branch |
| +1 | mvninstall | 582 | trunk passed |
| +1 | compile | 379 | trunk passed |
| +1 | checkstyle | 82 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 879 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 178 | trunk passed |
| 0 | spotbugs | 415 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 613 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 40 | Maven dependency ordering for patch |
| +1 | mvninstall | 541 | the patch passed |
| +1 | compile | 388 | the patch passed |
| +1 | javac | 388 | the patch passed |
| +1 | checkstyle | 86 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 686 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 174 | the patch passed |
| +1 | findbugs | 631 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 269 | hadoop-hdds in the patch passed. |
| -1 | unit | 1886 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 54 | The patch does not generate ASF License warnings. |
| | | 7747 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion |
|   | hadoop.ozone.om.TestSecureOzoneManager |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1361 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c5dc564ae7bd 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 915cbc9 |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/3/testReport/ |
| Max. process+thread count | 5395 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  public static final String SOURCE_NAME =
+      SCMPlacementMetrics.class.getSimpleName();
+  private static final MetricsInfo RECORD_INFO = Interns.info(SOURCE_NAME,
+      "SCM Placement Metrics");
+  private static MetricsRegistry registry;
+
+  // total datanode allocation request count
+  @Metric private MutableCounterLong datanodeRequestCount;
+  // datanode allocation tried count, including success, fallback and failed
+  @Metric private MutableCounterLong datanodeAllocationTryCount;
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java]: NIT: datanodeAllocationTryCount -> datanodeSelectAttemptCount

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  // datanode successful allocation count
+  @Metric private MutableCounterLong datanodeAllocationSuccessCount;
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java]: datanodeAllocationSuccessCount -> datanodeChooseSuccessCount

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  // datanode allocated with some allocation constrains compromised
+  @Metric private MutableCounterLong datanodeAllocationCompromiseCount;
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java]: datanodeAllocationCompromiseCount -> datanodeChooseFallbackCount

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  public SCMPlacementMetrics() {
+  }
+
+  public static SCMPlacementMetrics create() {
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMPlacementMetrics.java]: can we add a helper to unregister the metrics?

**Comment by ChenSammi** [comment]
Thanks Xiaoyu for the comments.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 41 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for branch |
| +1 | mvninstall | 588 | trunk passed |
| +1 | compile | 405 | trunk passed |
| +1 | checkstyle | 73 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 913 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 179 | trunk passed |
| 0 | spotbugs | 471 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 690 | trunk passed |
| -0 | patch | 521 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 41 | Maven dependency ordering for patch |
| +1 | mvninstall | 581 | the patch passed |
| +1 | compile | 406 | the patch passed |
| +1 | javac | 406 | the patch passed |
| +1 | checkstyle | 89 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 776 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 192 | the patch passed |
| +1 | findbugs | 689 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 286 | hadoop-hdds in the patch passed. |
| -1 | unit | 192 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 47 | The patch does not generate ASF License warnings. |
| | | 6408 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.ratis.TestOzoneManagerDoubleBufferWithOMResponse |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1361 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 897d688e963c 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 337e9b7 |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/4/testReport/ |
| Max. process+thread count | 1346 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1361/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
Thanks @ChenSammi  for the contribution. +1 for the latest change, I merged the change to trunk.

**Comment by ChenSammi** [comment]
Thanks @xiaoyuyao  for the review.

**Pull request merged**